ddl/ddl_api.go

```diff
@@ -2305,6 +2338,82 @@ func (d *ddl) AddColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTab
 	return errors.Trace(err)
 }
 
+// AddColumns will add multi new columns to the table.
+func (d *ddl) AddColumns(ctx sessionctx.Context, ti ast.Ident, specs []*ast.AlterTableSpec) error {
+	schema, t, err := d.getSchemaAndTableByIdent(ctx, ti)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Check all the columns at once.
+	addingColumnNames := make(map[string]bool)
+	dupColumnNames := make(map[string]bool)
+	for _, spec := range specs {
+		for _, specNewColumn := range spec.NewColumns {
+			if !addingColumnNames[specNewColumn.Name.Name.L] {
+				addingColumnNames[specNewColumn.Name.Name.L] = true
+				continue
+			}
+			if !spec.IfNotExists {
+				return errors.Trace(infoschema.ErrColumnExists.GenWithStackByArgs(specNewColumn.Name.Name.O))
+			}
+			dupColumnNames[specNewColumn.Name.Name.L] = true
+		}
+	}
+	columns := make([]*table.Column, len(addingColumnNames))
```

Comment:
Why not use `columns := make([]*table.Column, 0, len(addingColumnNames))`? Then we can remove line2393.

Reply:
```
// Since the newColumnsCount can be less than len(addingColumnNames)
columns = columns[:newColumnsCount]
```
`len(columns)`  can not determined in line 2363.

Reply:
You need to replace line `columns[newColumnsCount] = col` with `columns[newColumnsCount] = col`.